pkgs/development/compilers/circt/circt-llvm.nix

```diff
@@ -0,0 +1,50 @@
+{ stdenv
+, cmake
+, ninja
+, circt
+, llvm
+, python3
+}: stdenv.mkDerivation {
+  pname = circt.pname + "-llvm";
+  version = circt.version;
+  src = circt.src + /llvm;
+
+  requiredSystemFeatures = [ "big-parallel" ];
+
+  nativeBuildInputs = [ cmake ninja python3 ];
+
+  cmakeDir = "../llvm";
+  cmakeFlags = [
+    "-DBUILD_SHARED_LIBS=ON"
+    "-DLLVM_ENABLE_BINDINGS=OFF"
+    "-DLLVM_ENABLE_OCAMLDOC=OFF"
+    "-DLLVM_BUILD_EXAMPLES=OFF"
+    "-DLLVM_OPTIMIZED_TABLEGEN=ON"
+    "-DLLVM_ENABLE_PROJECTS=mlir"
+    "-DLLVM_TARGETS_TO_BUILD="
+    "-DLLVM_INSTALL_UTILS=ON"
+  ];
+
+  outputs = [ "out" "lib" "dev" ];
+
+  postInstall = ''
```

Comment:
It would be nice to have some comments explaining why there operations are necessary. For example (I guess) `llvm-config` is moved to `$dev` to avoid circular dependency.

This leads to another question: since circt-llvm is only used for compiling circt, is it necessary to split circt-llvm into multiple outputs, which brings these messy `postInstall` patches?

Reply:
> It would be nice to have some comments explaining why there operations are necessary. For example (I guess) llvm-config is moved to $dev to avoid circular dependency.

Good idea. I will work on it.

> This leads to another question: since circt-llvm is only used for compiling circt, is it necessary to split circt-llvm into multiple outputs, which brings these messy postInstall patches?

I did this so the resulting circt closure size doesn't get too big, since LLVM's output is quite large. 

The closure size without multiple outputs is: 390.44MiB
Same with multiple outputs is: 291.41MiB

I'll leave the diff I tried here.
<details>
  <summary>diff</summary>

```diff
diff --git a/pkgs/development/compilers/circt/circt-llvm.nix b/pkgs/development/compilers/circt/circt-llvm.nix
index 436f93b9c8f8..a37c3405e109 100644
--- a/pkgs/development/compilers/circt/circt-llvm.nix
+++ b/pkgs/development/compilers/circt/circt-llvm.nix
@@ -25,22 +25,6 @@
     "-DLLVM_INSTALL_UTILS=ON"
   ];
 
-  outputs = [ "out" "lib" "dev" ];
-
-  postInstall = ''
-    moveToOutput "bin/llvm-config*" "$dev"
-    moveToOutput "lib" "$lib"
-    moveToOutput "lib/cmake" "$dev"
-    substituteInPlace "$dev/lib/cmake/llvm/LLVMConfig.cmake" \
-      --replace 'set(LLVM_BINARY_DIR "''${LLVM_INSTALL_PREFIX}")' 'set(LLVM_BINARY_DIR "'"$lib"'")'
-    substituteInPlace \
-      "$dev/lib/cmake/llvm/LLVMExports-release.cmake" \
-      "$dev/lib/cmake/mlir/MLIRTargets-release.cmake" \
-      --replace "\''${_IMPORT_PREFIX}/lib/lib" "$lib/lib/lib" \
-      --replace "\''${_IMPORT_PREFIX}/lib/objects-Release" "$lib/lib/objects-Release" \
-      --replace "$out/bin/llvm-config" "$dev/bin/llvm-config"
-  '';
-
   doCheck = true;
   checkTarget = "check-mlir";
 
diff --git a/pkgs/development/compilers/circt/default.nix b/pkgs/development/compilers/circt/default.nix
index b7804f708228..dcf89b13f804 100644
--- a/pkgs/development/compilers/circt/default.nix
+++ b/pkgs/development/compilers/circt/default.nix
@@ -9,6 +9,7 @@
 , lit
 , gitUpdater
 , callPackage
+, autoPatchelfHook
 }:
 
 let
@@ -28,12 +29,12 @@ stdenv.mkDerivation rec {
 
   requiredSystemFeatures = [ "big-parallel" ];
 
-  nativeBuildInputs = [ cmake ninja git pythonEnv ];
+  nativeBuildInputs = [ cmake ninja git pythonEnv autoPatchelfHook ];
   buildInputs = [ circt-llvm ];
 
   cmakeFlags = [
     "-DBUILD_SHARED_LIBS=ON"
-    "-DMLIR_DIR=${circt-llvm.dev}/lib/cmake/mlir"
+    "-DMLIR_DIR=${circt-llvm}/lib/cmake/mlir"
 
     # LLVM_EXTERNAL_LIT is executed by python3, the wrapped bash script will not work
     "-DLLVM_EXTERNAL_LIT=${lit}/bin/.lit-wrapped"

```

</details>

Reply:
I see. Reduction of closure size does worth the effort.